### PR TITLE
Clear topicId cache when removing topic partitions

### LINKED ISSUE #22: Remove unused topic ids to avoid leaking memory
Topic IDs are added to [this map](https://github.com/satishd/kafka/blob/2.8.x-tiered-storage/core/src/main/scala/kafka/log/remote/RemoteLogManager.scala#L123-L124) but never cleaned up when a topic is deleted. This is needed in order to avoid a memory leak.

### PR DESCRIPTION
**Pull request by divijvaidya**
Resolves https://github.com/satishd/kafka/issues/22

## Problem
Topic IDs are added to `RemoteLogManager.scala` but never cleaned up when a topic is deleted. This may cause a memory leak.

## Change
This change clears topicId cache when removing topic partitions.

## Testing
1. Added a unit test to validate that the topics are indeed removed from the cache.
2. `./gradlew unitTest` is successful.

### CODE REVIEW

**Comment by divijvaidya** [comment]
@wyuka @satishd please take a look when you get a chance

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala [hunk 1]
```diff
@@ -120,8 +120,9 @@ class RemoteLogManager(fetchLog: TopicPartition => Option[Log],
   private val poolSize = rlmConfig.remoteLogManagerThreadPoolSize
   private val rlmScheduledThreadPool = new RLMScheduledThreadPool(poolSize)
 
-  // topic ids that are received on leadership changes, this map is NEVER cleared
-  private val topicIds: mutable.Map[String, Uuid] = mutable.Map.empty
+  // topic ids that are received on leadership changes, this map is cleared on stop partitions
+  private val topicPartitionIds: concurrent.Map[TopicPartition, Uuid] = new ConcurrentHashMap[TopicPartition, Uuid]().asScala
```
Comment (wyuka) [inline core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:125]: This seems a little wasteful. A single topic can have even thousands of partitions with replicas on a broker. We are duplicating the `topicId` UUID for each of these partitions.

Can we instead maintain a map of topicId to partitions, and keep the `topicIds` map? That way, when new partitions (leader or follower) are added, we can update the `topicIdToPartitionsMap`. When `stopPartitions`
 gets called with `delete = true`, we can remove from this map, and if the map becomes empty, we can remove from `topicIds`.

Comment (kamalcph) [reply core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:125]: The same UUID instance is reused for all the partitions. This approach lgtm.

**Comment by kamalcph** [comment]
LGTM

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala (hunk 1 above)
Comment (satishd) [reply core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:125]: UUIDs are just references and this code is easy to understand.

**Review by satishd — APPROVED** [review]
This is aligned with the changes introduced in the trunk [PR](https://github.com/apache/kafka/pull/11390). Thanks for backporting and adding more checks. LGTM.

**Comment by satishd** [comment]
@divijvaidya I will merge once the conflicts are resolved.

**Comment by divijvaidya** [comment]
@satishd this is ready for merge.

**Comment by satishd** [comment]
@divijvaidya Please resolve the conflicts.

**Comment by divijvaidya** [comment]
> @divijvaidya Please resolve the conflicts.

@satishd resolved the conflicts. Ready for review/merge.

**Comment by satishd** [comment]
Thanks @divijvaidya for resolving the conflicts.

**Pull request merged**